internal/errors.go

```diff
@@ -0,0 +1,80 @@
+package internal
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"go.temporal.io/sdk/temporal"
+)
+
+// errorType identifies a custom Cadence error as an AWS SDK error.
+const errorType = "AWS_SDK_ERROR"
+
+// EncodedError is a serialized version of the AWS SDK error.
+type EncodedError struct {
+	Code    string
+	Message string
+	OrigErr string
+}
+
+// Error implements the error interface.
+func (e EncodedError) Error() string {
+	return fmt.Sprintf("%s: %s", e.Code, e.Message)
+}
+
+// EncodeError encodes an error returned by the AWS SDK as a Cadence error.
+func EncodeError(err error) error {
+	if err == nil {
+		return nil
+	}
+
+	if aerr, ok := err.(awserr.Error); ok {
+		newErr := EncodedError{
+			Code:    aerr.Code(),
+			Message: aerr.Message(),
+		}
+
+		// TODO: try to encode the original error as well?
+		if aerr.OrigErr() != nil {
+			newErr.OrigErr = aerr.OrigErr().Error()
+		}
+
+		return temporal.NewApplicationError(newErr.Error(), errorType, newErr)
+	}
+
+	return err
+}
+
+// DecodedError is a serialized version of the AWS SDK error.
+type DecodedError struct {
+	Code    string
+	Message string
+	OrigErr error
+}
+
+// DecodedError implements the error interface.
+func (e DecodedError) Error() string {
+	return fmt.Sprintf("%s: %s", e.Code, e.Message)
+}
+
+// DecodeError decodes an error returned by a Cadence activity.
+// If the error is an AWS SDK error, it gets converted to an AWS SDK compatible error interface.
+// Otherwise the original error is returned.
```

Comment:
I would not decode error. It loses information about the activity (like activityId, etc) and doesn't follow pattern other activities use. I would provide a helper method to extract code from it. Something like
```
clients.GetErrorCode(err error) string
```

Reply:
Makes sense. Where should it go though? In the clients package? A separate errors (or similar) package sounds better to me. Or an overall helper package. In any case, this is not generated code, so the clients package doesn't feel right.

Reply:
I'd probably pair it with a `IsAWSError` function as well.

Reply:
Not sure about the name. Maybe tawssdk?